arch/x86/core/ia32/x86_mmu.c

```diff
@@ -224,6 +244,162 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr,
 	}
 }
 
+static char __aligned(MMU_PAGE_SIZE)
+	page_pool[MMU_PAGE_SIZE * CONFIG_X86_MMU_PAGE_POOL_PAGES];
+
+static char *page_pos = page_pool + sizeof(page_pool);
+
+static void *get_page(void)
+{
+	page_pos -= MMU_PAGE_SIZE;
+
+	__ASSERT(page_pos >= page_pool, "out of MMU pages\n");
+
+	return page_pos;
+}
+
+__aligned(0x20) struct x86_mmu_pdpt z_x86_kernel_pdpt;
+#ifdef CONFIG_X86_KPTI
+__aligned(0x20) struct x86_mmu_pdpt z_x86_user_pdpt;
+#endif
+
+extern char z_shared_kernel_page_start[];
+
+static void add_mmu_region_page(struct x86_mmu_pdpt *pdpt, uintptr_t addr,
+				u64_t flags, bool user_table)
+{
+	union x86_mmu_pdpte *pdpte;
+	struct x86_mmu_pd *pd;
+	union x86_mmu_pde_pt *pde;
+	struct x86_mmu_pt *pt;
+	union x86_mmu_pte *pte;
+
+	/* Setup the PDPTE entry for the address, creating a page directory
+	 * if one didn't exist
+	 */
+	pdpte = X86_MMU_GET_PDPTE(pdpt, addr);
+	if (pdpte->p == 0) {
+		pd = get_page();
+		pdpte->pd = ((uintptr_t)pd) >> MMU_PAGE_SHIFT;
+	} else {
+		pd = (struct x86_mmu_pd *)(pdpte->pd << MMU_PAGE_SHIFT);
+	}
+	pdpte->value |= (flags & PDPTE_FLAGS_MASK);
+
+	/* Setup the PDE entry for the address, creating a page table
+	 * if necessary
+	 */
+	pde = X86_MMU_GET_PDE(pdpt, addr);
+	if (pde->p == 0) {
+		pt = get_page();
+		pde->pt = ((uintptr_t)pt) >> MMU_PAGE_SHIFT;
+	} else {
+		pt = (struct x86_mmu_pt *)(pde->pt << MMU_PAGE_SHIFT);
+	}
+	pde->value |= (flags & PDE_FLAGS_MASK);
+	/* Execute disable bit needs special handling, we should only set it
+	 * at the page directory level if ALL pages have XD set (instead of
+	 * just one).
+	 *
+	 * Use ignored1 field as a marker, its for OS use anyway, it's
+	 * the 'dirty' bit.
```

Comment:
need to check if more flags need handling in this way.

use an unused bit position for the OS that doesn't get touched by the cpu